File: dsa/strings_coursera/1_suffix_trees/assignment/trie_matching_extended/trie_matching_extended.go
package main

import "fmt"

type node struct {
	val        int
	endPattern bool
	edges      map[rune]*node
}

func (n *node) String() string {
	return fmt.Sprintf("val: %d, endPattern: %v, edges: %v", n.val, n.endPattern, n.edges)
}

func buildTrie(patterns []string) *node {
	root := &node{
		val:        0,
		endPattern: false,
		edges:      make(map[rune]*node),
	}
	nc := 0
	for _, pattern := range patterns {
		current := root
		for idx, c := range pattern {
			j, ok := current.edges[c]
			if ok {
				if !j.endPattern {
					fmt.Printf("Edge found for %v, isEnd, %v\n", c, idx == len(pattern)-1)
					j.endPattern = idx == len(pattern)-1
				}
				current = j
			} else {
				nc++
				n := &node{
					val:        nc,
					endPattern: idx == len(pattern)-1,
					edges:      make(map[rune]*node),
				}
				current.edges[c] = n
				current = n
			}
		}
	}
	return root
}

func match(text string, patterns []string) []int {
	ans := make([]int, 0)
	trie := buildTrie(patterns)
	fmt.Printf("%v\n", trie)
	h := 0
	for h < len(text) {
		if prefixTrieMatch(text[h:], trie) {
			ans = append(ans, h)
		}
		h++
	}
	return ans

}

func prefixTrieMatch(s string, trie *node) bool {
	f := 0
	root := trie

	for {
		// This is a leaf
		if root.endPattern {
			return true
		} else {
			if len(s) <= f {
				return false
			} else if w, ok := root.edges[rune(s[f])]; ok {
				f++
				root = w
			} else {
				return false
			}
		}
	}
}

func main() {
	fmt.Println(match("ACATA", []string{"AT", "A", "AG"}))
}

File: dsa/strings_coursera/1_suffix_trees/assignment/trie_matching_extended/TrieMatchingExtended.java
import java.io.*;
import java.util.*;

class Node
{
	public static final int Letters =  4;
	public static final int NA      = -1;
	public int next [];
	public boolean patternEnd;

	Node ()
	{
		next = new int [Letters];
		Arrays.fill (next, NA);
		patternEnd = false;
	}

	@Override
	public String toString(){
		StringBuilder sb = new StringBuilder();
		sb.append(patternEnd);
		sb.append("(");
		for (int i = 0; i < next.length; i++) {
			sb.append(next[i]);	
			sb.append(" ");	
		}
		sb.append(")");
		return sb.toString();
	}
}

public class TrieMatchingExtended implements Runnable {
	int letterToIndex (char letter)
	{
		switch (letter)
		{
			case 'A': return 0;
			case 'C': return 1;
			case 'G': return 2;
			case 'T': return 3;
			default: assert (false); return Node.NA;
		}
	}

	List<Node> buildTrie(List<String> patterns) {
        List<Node> trie = new ArrayList<>();
        trie.add(new Node());
        for(String pattern : patterns){
            Node current = trie.get(0);
            for(int i=0; i < pattern.length(); i++){
				Character c = pattern.charAt(i);
				int idx = letterToIndex(c);
                int val = current.next[idx];
                if(val != Node.NA){
					Node node = trie.get(val);
					if (!node.patternEnd){
						node.patternEnd = i == pattern.length() - 1;
					}
                    current = node;
                }else{
                    current.next[idx] = trie.size();
					Node n = new Node();
					n.patternEnd = i == pattern.length() - 1 ;
                    trie.add(n);
					current = n;
                }
            }
			// System.out.println(trie);
        }
        return trie;
    }

	List <Integer> solve (String text, int n, List <String> patterns) {
		List <Integer> result = new ArrayList <Integer> ();
		List<Node> trie = buildTrie(patterns);

		int h = 0;
		while(h < text.length()){
			if(prefixTrieMatch(text.substring(h), trie)){
				result.add(h);
			}
			h++;
		}
		return result;
	}

	boolean prefixTrieMatch(String text, List<Node> trie){
		int f = 0;
		int v = 0;

		while(true) {
			if(trie.get(v).patternEnd){
				return true;
			}else{
				if(text.length() <= f){
					return false;
				}else{
					Node node = trie.get(v);
					char c = text.charAt(f);
					if(node.next[letterToIndex(c)] != Node.NA){
						v = node.next[letterToIndex(c)];
						f++;
					}else{
						return false;
					}
				}
			}
		}
	}

	public void run () {
		try {
			BufferedReader in = new BufferedReader (new InputStreamReader (System.in));
			String text = in.readLine ();
		 	int n = Integer.parseInt (in.readLine ());
		 	List <String> patterns = new ArrayList <String> ();
			for (int i = 0; i < n; i++) {
				patterns.add (in.readLine ());
			}

			List <Integer> ans = solve (text, n, patterns);

			for (int j = 0; j < ans.size (); j++) {
				System.out.print ("" + ans.get (j));
				System.out.print (j + 1 < ans.size () ? " " : "\n");
			}
		}
		catch (Throwable e) {
			e.printStackTrace ();
			System.exit (1);
		}
	}

	public static void main (String [] args) {
		new Thread (new TrieMatchingExtended ()).start();
	}
}
